csharp/src/Google.Protobuf/Reflection/CustomOptions.cs

```diff
@@ -205,98 +232,116 @@ internal CustomOptions(IDictionary<int, IExtensionValue> values)
+                {
+                    if (extensionValue is ExtensionValue<T>)
+                    {
+                        ExtensionValue<T> single = extensionValue as ExtensionValue<T>;
+                        ByteString bytes = single.GetValue().ToByteString();
+                        value = new T();
+                        value.MergeFrom(bytes);
+                        return true;
+                    }
+                    else if (extensionValue is RepeatedExtensionValue<T>)
+                    {
+                        RepeatedExtensionValue<T> repeated = extensionValue as RepeatedExtensionValue<T>;
+                        value = repeated.GetValue()
+                            .Select(v => v.ToByteString())
+                            .Aggregate(new T(), (t, b) =>
+                            {
+                                t.MergeFrom(b);
+                                return t;
+                            });
+                        return true;
+                    }
+                }
             }
 
-            IExtensionValue extensionValue;
-            if (values.TryGetValue(field, out extensionValue))
+            UnknownField f;
+            if (UnknownFieldSet.TryGetField(unknownFields, field, out f))
             {
-                if (extensionValue is ExtensionValue<T>)
+                value = new T();
+                foreach (var bytes in f.LengthDelimitedList ?? Enumerable.Empty<ByteString>())
                 {
-                    ExtensionValue<T> single = extensionValue as ExtensionValue<T>;
-                    ByteString bytes = single.GetValue().ToByteString();
-                    value = new T();
                     value.MergeFrom(bytes);
-                    return true;
-                }
-                else if (extensionValue is RepeatedExtensionValue<T>)
-                {
-                    RepeatedExtensionValue<T> repeated = extensionValue as RepeatedExtensionValue<T>;
-                    value = repeated.GetValue()
-                        .Select(v => v.ToByteString())
-                        .Aggregate(new T(), (t, b) =>
-                        {
-                            t.MergeFrom(b);
-                            return t;
-                        });
-                    return true;
                 }
+                return true;
             }
 
             value = null;
             return false;
         }
 
-        private bool TryGetPrimitiveValue<T>(int field, out T value)
-        {
-            if (values == null)
-            {
-                value = default(T);
-                return false;
-            }
+        private delegate bool TryGetLastValue<T>(UnknownField field, out T value);
 
-            IExtensionValue extensionValue;
-            if (values.TryGetValue(field, out extensionValue))
+        private bool TryGetPrimitiveValue<T, U>(int field, TryGetLastValue<U> getValue, Func<U, T> conv, out T value)
```

Comment:
The meaning of the `U` generic argument isn't clear at all from the first glance (I'm guessing it's the type of the result for the unknown field, but it's very unobvious).
Perhaps the refactoring I proposed should help with that.

Btw, the `U` argument also looks useless. Why cannot the getValueFunc get the value AND perform the conversion, eliminating the need for `U` completely?

Reply:
The way it's designed currently is optimized to remove code duplication. Since most values in the UnknownField are represented different from how they're exposed in `TryGet*` methods, I abstracted the `TryGetLast*` out and made it so you just need to provide the value's custom conversion method. Making it so the `getValueFunc` had to get the value and perform the conversion would increase the code size tremendously. Plus anonymous delegates can't have `out` parameters. So we'd have to use tuples to communicate if a value existed in the unknown field and the converted value, unless we want to make a separate local function just to do it.